### PR TITLE
BSRD-787 Add `migrate.analyze-addons` task

### PR DESCRIPTION
**Pull request by sebalix**
Port of https://github.com/camptocamp/cosanum_odoo/pull/2307, already tested on 2-3 projects.

Would need some improvements, like listing paid modules in the report as it's done for local ones, but that's something that can be done later.

**TODO:**

- [ ] switch to Excel instead of CSV (to use tabs)
- [ ] list server actions
- [ ] list studio fields
- [ ] list studio views
- [ ] list mail templates

### CODE REVIEW

File: odoo_tools/tasks/main.py [hunk 1]
```diff
@@ -1,11 +1,12 @@
 from invoke import Collection, Program
 
-from . import database, lastpass, module, pr, translate
+from . import database, lastpass, migrate, module, pr, translate
 
 tasks_ns = Collection()
 
 tasks_ns.add_collection(database)
 tasks_ns.add_collection(lastpass)
+tasks_ns.add_collection(migrate)
```
Comment (simahawk) [inline odoo_tools/tasks/main.py:9]: New features should be added as new cli tools, not tasks.

File: odoo_tools/tasks/migrate.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+import git
+import oca_port
+from invoke import task
+
+from ..utils.path import cd
+from .database import get_db_request_result
+
+ANALYZE_SUBMODULES_TO_SKIP = []
+ANALYZE_SUBMODULES = {
```
Comment (simahawk) [inline odoo_tools/tasks/migrate.py:18]: This should be made configurable. Easy option: accept an INI file as cmd line arg.
In the utils you'll find some helper to read ini files.
Or go for a yaml file if you prefer.

File: odoo_tools/tasks/migrate.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+class AnalyzeAddons:
+    """Get a migration report of modules.
+
+    Modules belonging to `odoo/local-src` will be listed as `to migrate`,
+    while OCA modules will be analyzed by `oca-port` PRs to check if they
+    could to be migrated, or if some commits/PRs could be ported.
+    """
+
+    def __init__(self, ctx, from_branch, to_branch, database_name):
+        self.ctx = ctx
```
Comment (simahawk) [inline odoo_tools/tasks/migrate.py:39]: AFAIS this is needed only to use `get_db_request_result` but that requires refactoring of the db tasks.
IMO you should accept a DSN or even an open connection to the DB. That's it.

File: odoo_tools/tasks/migrate.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        self.from_branch = from_branch
+        self.to_branch = to_branch
+        self.database_name = database_name
+        if not os.environ.get("GITHUB_TOKEN"):
+            sys.exit(
```
Comment (simahawk) [inline odoo_tools/tasks/migrate.py:44]: pls use `ui.exit_msg`

File: odoo_tools/tasks/migrate.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                "Please set your GitHub token in the GITHUB_TOKEN environment variable."
+            )
+
+    def _get_installed_local_addons(self):
+        # Get the local addons
+        local_addons = [
+            os.path.basename(addon)
+            for addon in os.scandir("./odoo/local-src/")
```
Comment (simahawk) [inline odoo_tools/tasks/migrate.py:52]: use `get_conf_key("local_src_rel_path")`  to get the path for the project.

File: odoo_tools/tasks/migrate.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        )
+        return [
+            row[0] for row in get_db_request_result(self.ctx, self.database_name, sql)
+        ]
+
+    def _get_all_external_addons(self):
+        """Return all addons located in `odoo/external-src/`."""
+        external_addons = {}
+        external_addons_by_repo = collections.defaultdict(list)
+        for repo in os.scandir("./odoo/external-src/"):
```
Comment (simahawk) [inline odoo_tools/tasks/migrate.py:72]: get_conf_key

Comment (simahawk) [reply odoo_tools/tasks/migrate.py:72]: actually I'd say these paths should be passed as params to the class

**Review by simahawk — CHANGES_REQUESTED** [review]

File: odoo_tools/tasks/main.py (hunk 1 above)
Comment (sebalix) [reply odoo_tools/tasks/main.py:9]: Yes that's what I guessed by looking here and there, but it was the first step to get something working ASAP by copying the former invoke task.

File: odoo_tools/tasks/migrate.py (hunk 4 above)
Comment (sebalix) [reply odoo_tools/tasks/migrate.py]: done, for now it also prints the traceback but I guess it won't be the case once the task is migrated to use `click`.

**Comment by sebalix** [comment]
Closing as this task is now handled by another internal tool.

**Pull request closed without merging**